### PR TITLE
fix(developer): Debug character grid performance

### PR DESCRIPTION
**Pull request by mcdurdin**
The character grid in the debugger would become very slow as the number of characters increased, with substantial flickering. This refactor removes the unnecessary recalculations of the grid cell count and thus resolves performance problems.

### CODE REVIEW

**Comment by jahorton** [comment]
This will probably need some user testing to be safe; the code changes are _very_ significant and hard to follow, as I can't follow the code changes.  It looks like many of the calcuation statements were completely restructured.

**Comment by mcdurdin** [comment]
# User Testing

1. Open Keyman Developer. load a keyboard, press F5 to start debugging.
2. Type various strings and ensure that the displayed character cells at the bottom of the debug window are correct.
3. Type enough text so that it doesn't fit in the displayed character cells. Ensure that the cells shown correspond to the characters to the left of the cursor position.
4. Select some text in the debugger and check that the cells match the selected text.
5. Run these tests again with a keyboard that includes non-BMP (i.e. > U+FFFF) characters and/or deadkeys.

@MakaraSok

**Comment by MakaraSok** [comment]
Tested and it works as expected. 
- the cells shown do correspond to the characters to the left of the cursor position
- the cells do match the selected text when move the cursor around, even with deadkeys

**Review by jahorton — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.228-beta](https://keyman.com/downloads/releases/beta/14.0.228)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.9-alpha](https://keyman.com/downloads/releases/alpha/15.0.9)
